File: solutions/javascript/anagram/1/anagram.js
//
// This is only a SKELETON file for the 'Anagram' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const findAnagrams = (word, candidates) => {
  word = word.toLowerCase();
  const set = [];
  for(let candidate of candidates) {
    const candidateLower = candidate.toLowerCase();
    if(candidateLower != word && candidate.length == word.length && isAnagram(candidateLower, word))
      set.push(candidate);
  }
  return set;
};

const isAnagram = (w1, w2) => {
  const frecuencies = {};
  for(const char of w1)
    frecuencies[char] = (frecuencies[char] || 0) + 1;
  for(const char of w2){
    if (!frecuencies[char]) return false;
    frecuencies[char]--;
  }
  return Object.values(frecuencies).every(count => count === 0);
}

File: solutions/python/anagram/1/anagram.py
def find_anagrams(word, candidates):
    word = word.lower()
    result = []
    find_all_combinations("", result, word)
    anagrams = []
    for candidate in candidates:
        if candidate.lower() != word and candidate.lower() in result:
            anagrams.append(candidate)
    return anagrams

def find_all_combinations(prefix, result, word):
    if len(word) == 0:
        result.append(prefix)
        return

    for char in word:
        find_all_combinations(prefix + char, result, word.replace(char, '', 1))
